## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [427]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0    1,096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [428]:
# View the column names in the data
df.columns



Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [429]:
# Use the describe function to gather some basic statistics
df.describe()


client_id     order_id  order_week  order_year  unit_price  unit_cost  \
count  54,639.00    54,639.00   54,639.00   54,639.00   54,639.00  54,639.00   
mean   54,837.87 5,470,190.03       11.36    2,022.99      136.27      99.45   
std    25,487.44 2,599,806.93        7.02        0.08      183.87     133.16   
min    10,033.00 1,000,886.00        1.00    2,022.00        0.01       0.01   
25%    33,593.00 3,196,372.00        6.00    2,023.00       20.80      14.84   
50%    53,305.00 5,496,966.00       11.00    2,023.00       68.31      49.89   
75%    78,498.00 7,733,869.00       17.00    2,023.00      173.16     125.57   
max    99,984.00 9,998,480.00       52.00    2,023.00    1,396.23     846.27   

       unit_weight          qty  line_number  
count    54,639.00    54,639.00    54,639.00  
mean          5.00       570.26         2.98  
std           5.33    18,795.52         2.44  
min           0.00         0.00         0.00  
25%           1.45        32.00         1.00  
50%           3.24        68.00         3.00  
75%           6.89       170.00         5.00  
max          46.43 3,958,244.00         9.00

In [430]:
# Use this space to do any additional research
# and familiarize yourself with the data.


In [431]:
# What three item categories had the most entries?
top_categories = df['category'].value_counts().head(3)
top_categories


category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [432]:
# For the category with the most entries,
# which subcategory had the most entries?
category_filter = df[df['category'] == 'consumables']
# category_filter.head()
# counts = category_filter['subcategory'].value_counts()
# counts.head()
top_subcategory = category_filter['subcategory'].value_counts().idxmax()
print(f"The top subcategory for the category with the most entries (consumables) is : {top_subcategory}")

The top subcategory for the category with the most entries (consumables) is : bathroom supplies


In [433]:
# Which five clients had the most entries in the data?
most_entries = df[['first','last','client_id']].value_counts().head(5)
print(f"The 5 clients with the most entries in the data are")
most_entries




The 5 clients with the most entries in the data are


first      last     client_id
Jessica    Reyes    33615        220
Angela     Everett  66037        211
Bryan      Myers    46820        209
Kendra     Garrett  24741        207
Alexandra  Young    38378        207
Name: count, dtype: int64

In [434]:
# Store the client ids of those top 5 clients in a list.
client_id_list = df['client_id'].value_counts().head(5).index.tolist()
client_id_list


[33615, 66037, 46820, 38378, 24741]

In [435]:
# How many total units (the qty column) did the
# client with the most entries order order?
total_units = df.loc[df['client_id'] == 33615]['qty'].sum()
# total_units
print(f"The client with the most entries (Jessica Reyes) ordered {total_units} units")


The client with the most entries (Jessica Reyes) ordered 64313 units


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [436]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty
df['subtotal'] = df['unit_price']*df['qty']
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number   subtotal  
0    1,096.80     762.71         7.50  105            1 115,164.00  
1       24.95      15.09         1.49   21            0     523.95  
2       13.52       7.86         1.68   39            6     527.28  
3       36.42      24.85         1.23   29            3   1,056.18  
4      195.10     108.17        46.43   20            1   3,902.00

In [437]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.

df['shipping_price'] = df.apply(lambda row: 7 * row['unit_weight'] * row['qty'] if (row['unit_weight'] * row['qty']) > 50 else 10 * row['unit_weight'] * row['qty'], axis=1)
df.head(10)

first      last                                           job  \
0     Donald   Harding                                  Immunologist   
1    Tiffany     Myers                               Music therapist   
2    Shannon    Watson                                  Immunologist   
3     Nathan     Baker                         Accounting technician   
4  Christina  Schwartz                                  Chiropractor   
5    Jessica     Reyes  Production designer, theatre/television/film   
6      Brian     Moore                                       Barista   
7      Jesus    Wright                            Theme park manager   
8    Rebecca     Brown                                   Chiropodist   
9       Greg    Garcia                            Secretary, company   

                 phone                                email  client_id  \
0   793-904-7725x39308     harding.donald.7185@sullivan.com      58515   
1     201.442.4543x942    myers.t.6537@ferguson-johnson.net      37609   
2    687.737.9424x8503                swatson8146@payne.net      57113   
3     827-788-8123x012               bakernathan@benson.com      46554   
4         265-829-3643  christinaschwartz9252@mcconnell.com      92089   
5           5016915923             jessica.reyes@grimes.net      33615   
6  +1-893-400-2733x982          moore.official@martinez.com      87213   
7   605.947.9467x66513        wright.jesus.7579@meadows.com      46709   
8     665-955-0156x926                 rbrown850@lewis.info      73362   
9   (231)342-7216x1421                  ggarcia@bentley.biz      58789   

   order_id  order_date  order_week  order_year         item_id     category  \
0   8953482  2023-04-28          17        2023  EUD29711-63-6U        decor   
1   8069089  2023-05-19          20        2023  XDA18116-89-4A  consumables   
2   1902144  2023-01-29           4        2023  ABE59463-05-7E     software   
3   9031802  2023-04-25          17        2023  ZMM00836-65-0C  consumables   
4   1322274  2023-05-28          21        2023  BZX55559-12-3X  consumables   
5   1327730  2023-05-02          18        2023  BMB23158-88-9A        decor   
6   8080280  2023-04-19          16        2023  UAC50632-25-4E  consumables   
7   8946782  2023-02-14           7        2023  UEU25316-01-1U  consumables   
8   9394118  2023-05-13          19        2023  DDA06248-98-7Z  consumables   
9   9665852  2023-04-14          15        2023  DUU85042-87-5E    furniture   

          subcategory  unit_price  unit_cost  unit_weight  qty  line_number  \
0            wall art    1,096.80     762.71         7.50  105            1   
1                pens       24.95      15.09         1.49   21            0   
2  project management       13.52       7.86         1.68   39            6   
3                pens       36.42      24.85         1.23   29            3   
4                misc      195.10     108.17        46.43   20            1   
5              plants      172.84     154.44         2.79   78            0   
6    kitchen supplies      245.56     231.95         1.73   56            5   
7                pens       20.99      16.97         5.12   22            4   
8      paper products       83.86      59.00         0.27   49            3   
9             seating       66.88      54.15         2.47   78            1   

    subtotal  shipping_price  
0 115,164.00        5,512.50  
1     523.95          312.90  
2     527.28          458.64  
3   1,056.18          356.70  
4   3,902.00        6,500.20  
5  13,481.52        1,523.34  
6  13,751.36          678.16  
7     461.78          788.48  
8   4,109.14          132.30  
9   5,216.64        1,348.62

In [438]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%

df['line_price'] = df['subtotal'] + df['shipping_price'] + ((df['subtotal'] + df['shipping_price']) * 0.0925)
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...     category         subcategory unit_price  \
0          17        2023  ...        decor            wall art   1,096.80   
1          20        2023  ...  consumables                pens      24.95   
2           4        2023  ...     software  project management      13.52   
3          17        2023  ...  consumables                pens      36.42   
4          21        2023  ...  consumables                misc     195.10   

   unit_cost  unit_weight  qty  line_number   subtotal  shipping_price  \
0     762.71         7.50  105            1 115,164.00        5,512.50   
1      15.09         1.49   21            0     523.95          312.90   
2       7.86         1.68   39            6     527.28          458.64   
3      24.85         1.23   29            3   1,056.18          356.70   
4     108.17        46.43   20            1   3,902.00        6,500.20   

   line_price  
0  131,839.08  
1      914.26  
2    1,077.12  
3    1,543.57  
4   11,364.40  

[5 rows x 21 columns]

In [439]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).

df['cost'] = df['unit_cost']*df['qty'] + df['shipping_price']
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...         subcategory unit_price unit_cost  \
0          17        2023  ...            wall art   1,096.80    762.71   
1          20        2023  ...                pens      24.95     15.09   
2           4        2023  ...  project management      13.52      7.86   
3          17        2023  ...                pens      36.42     24.85   
4          21        2023  ...                misc     195.10    108.17   

   unit_weight  qty  line_number   subtotal  shipping_price  line_price  \
0         7.50  105            1 115,164.00        5,512.50  131,839.08   
1         1.49   21            0     523.95          312.90      914.26   
2         1.68   39            6     527.28          458.64    1,077.12   
3         1.23   29            3   1,056.18          356.70    1,543.57   
4        46.43   20            1   3,902.00        6,500.20   11,364.40   

       cost  
0 85,597.05  
1    629.79  
2    765.18  
3  1,077.35  
4  8,663.60  

[5 rows x 22 columns]

In [440]:
# Create a column for the profit of
# each line using line cost and line price
df['profit'] = df['line_price'] - df['cost']
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...   1,096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   

   line_number   subtotal  shipping_price  line_price      cost    profit  
0            1 115,164.00        5,512.50  131,839.08 85,597.05 46,242.03  
1            0     523.95          312.90      914.26    629.79    284.47  
2            6     527.28          458.64    1,077.12    765.18    311.94  
3            3   1,056.18          356.70    1,543.57  1,077.35    466.22  
4            1   3,902.00        6,500.20   11,364.40  8,663.60  2,700.80  

[5 rows x 23 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [441]:
# Check your work using the totals above
order_totals = df.groupby('order_id')['line_price'].sum()

specified_order_ids = [2742071, 2173913, 6128929]

specified_order_totals = order_totals.loc[specified_order_ids]

specified_order_totals

order_id
2742071   152,811.90
2173913   162,388.72
6128929   923,441.24
Name: line_price, dtype: float64

## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [442]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.
pd.options.display.float_format = '{:,.2f}'.format
top_5_spending = df[df['client_id'].isin(client_id_list)].groupby('client_id')['line_price'].sum()
top_5_spending.head()


client_id
24741   82,268,891.98
33615    8,377,308.57
38378   12,906,550.88
46820    9,743,794.32
66037   10,259,514.80
Name: line_price, dtype: float64

In [443]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.
summary_df = df[df['client_id'].isin(client_id_list)].groupby('client_id').agg({'qty':'sum', 'shipping_price':'sum', 'line_price':'sum', 'profit':'sum'}).sort_values('profit', ascending=False)
summary_df.head()


qty  shipping_price    line_price        profit
client_id                                                    
24741      239862    5,126,448.37 82,268,891.98 36,579,992.27
38378       73667    3,429,455.40 12,906,550.88  3,271,829.90
66037       43018    1,395,151.85 10,259,514.80  3,255,031.82
46820       75768    1,601,448.84  9,743,794.32  2,736,602.68
33615       64313    1,828,984.89  8,377,308.57  2,201,994.66

In [444]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.

def convert_to_millions(column):
    return (column/1000000)


summary_df[['shipping_price', 'line_price', 'profit']] = summary_df[['shipping_price', 'line_price', 'profit']].apply(convert_to_millions)

names = { 'qty': 'Total Units Purchased', 'shipping_price': 'Total Shipping Price (Millions of Dollars)', 'line_price': 'Total Revenue (Millions of Dollars)', 'profit': 'Total Profit (Millions of Dollars)'}
summary_df.rename(columns=names, inplace=True)
summary_df.head()


Total Units Purchased  Total Shipping Price (Millions of Dollars)  \
client_id                                                                      
24741                     239862                                        5.13   
38378                      73667                                        3.43   
66037                      43018                                        1.40   
46820                      75768                                        1.60   
33615                      64313                                        1.83   

           Total Revenue (Millions of Dollars)  \
client_id                                        
24741                                    82.27   
38378                                    12.91   
66037                                    10.26   
46820                                     9.74   
33615                                     8.38   

           Total Profit (Millions of Dollars)  
client_id                                      
24741                                   36.58  
38378                                    3.27  
66037                                    3.26  
46820                                    2.74  
33615                                    2.20

In [445]:
# Sort the updated data by "Total Profit" form highest to lowest
summary_df.sort_values('Total Profit (Millions of Dollars)', ascending=False, inplace=True)
summary_df.head()

Total Units Purchased  Total Shipping Price (Millions of Dollars)  \
client_id                                                                      
24741                     239862                                        5.13   
38378                      73667                                        3.43   
66037                      43018                                        1.40   
46820                      75768                                        1.60   
33615                      64313                                        1.83   

           Total Revenue (Millions of Dollars)  \
client_id                                        
24741                                    82.27   
38378                                    12.91   
66037                                    10.26   
46820                                     9.74   
33615                                     8.38   

           Total Profit (Millions of Dollars)  
client_id                                      
24741                                   36.58  
38378                                    3.27  
66037                                    3.26  
46820                                    2.74  
33615                                    2.20

### Summary
Client 24741 produces the most revenue for the company by far, at 36.58 million, while no other client had more than 3.27 million in revenue. This is likely because that client also had almost 3 times as many units purchesed when compared to the client with the next most units. Client 66037 orderecd signifigantly less units than other clients, but still had the 3rd most revenue at 3.26 million, indicating they are buying more expensive units. 